### ISSUE TITLE
default benchmark ranking is bad. seriously.

### ISSUE THREAD

**Issue by vanhauser-thc** [google/fuzzbench#366]
In https://www.fuzzbench.com/reports/2020-05-20-aflplusplus-2/index.html different afl++ variants were run.
The overall ranking is: default (simple aflplusplus), cmplog, laf, mopt

However when we measure the overall total blocks found:

|fuzzer|total bbs|
|----|----|
|aflplusplus_instrim|69899.0|
|aflplusplus|70521.5|
|aflplusplus_mopt2|70357.0|
|aflplusplus_mopt|69426.5|
|aflplusplus_cmplog|70817.5|
|aflplusplus_laf|71393.0|
|aflplusplus_laf_17|71147.0|

the ranking is: laf, laf17, cmplog, default, ...

When we look at % of maximum coverage achieved per target (so the best always receives 1.0 and the others less) and add them up:

|fuzzer|total % of bb per target|
|----|----|
|aflplusplus_instrim|19.095|
|aflplusplus|19.357|
|aflplusplus_mopt2|19.241|
|aflplusplus_mopt|18.918|
|aflplusplus_cmplog|20.248|
|aflplusplus_laf|20.339|
|aflplusplus_laf_17|20.124|

the ranking is: laf, cmplog, laf17, default, ...

IMHO the second ranking method is better than the first, however in this benchmark run they come to  similar ranking order.

If I compare that with the report ranks:
 * default (simple aflplusplus) is displayed the best, in both alternative rankings it is 4th place
 * mopt is in both alternative rankings the last place, in the report it is in the middle (4th place)
 * the best in the alternative rankings merely makes rank 3 in the report

Because of this  I always do my assessment in an excel sheet that I auto-generate from a copy-paste from the median coverages table in the report.

TLDR - what I tried to show is that the ranking method in the reports give a false impression what is good and what is bad.
IMHO that needs to be fixed.

**Comment by wideglide** [comment]
I should probably understand @mboehme 's paper a bit better before commenting, because he would argue that we should also rank fuzzers based on their time to reach coverage X (#327) instead of coverage improvement in a fixed time budget.

But that still doesn't account for how to rank fuzzers based on performance across N number of benchmarks.  Should each benchmark receive equal weighting in the ranking?  For instance, how do you compare fairly when cmplog achieves better coverage on several benchmarks, but significantly worse coverage on sqlite3, vorbis and woff2?

The great thing about the Fuzzbench analysis module is that you can easily use the other ranking algorithms to see how they compare. (see this [Google colab analysis](https://colab.research.google.com/drive/1-_6oUD4Xz3Rm9LXpgNbns6vwteW5_hbL))  I think the _average normalized score_ ranking is the closest to your % maximum coverage ranking above.  The default (simple aflplusplus) achieves the best (unweighted) average ranking across all benchmarks, the second most "wins", and the best ranking of stats test wins.

I also did a quick [spreadsheet analysis](https://docs.google.com/spreadsheets/d/1pYaf4yrLj0VM2wicfQnfF0A0d_H3KIqukIuiLJPEaNY/edit?usp=sharing) for comparison.

**Comment by vanhauser-thc** [comment]
@wideglide 

I havent read his paper, but that would also be a good method to measure fuzzers IMHO.

The problem with the the current method is that there are some test cases where the difference of the fuzzers is between 0.5 and 2 basic blocks (e.g. jsoncpp). and then there are others where there are very large differences (e.g. libpcap). And a fuzzer that is good on the first ones, but very bad on the last ones will be rated a good fuzzer. Which I think most people would agree is not the correct conclusion. (for the first type there is not an effective difference between the fuzzers, for the the last type it is).

Another option would be to remove the benchmarks where there is only a minimal difference between the fuzzers as these are rather pointless and just cost CPU time to perform.

I know that there are other algorithms and that I can generate my own reports. but that was not my point. my point was the goal of the project is to show which fuzzers perform better than others - I mean that is the first and main chart being displayed. and I want to point out that this ranking algorithm basically gives the wrong advice to people who want to choose a good fuzzer (without the need to download the stats and generate a report for themselves and who might not have the necessary background to know which analysis type to choose ...)

I havent run the numbers (and the afl++ ones cant be compared because they were run differently), but my guess is that it would further propel honggfuzz and afl-smart even more to the front, laf should also move up, and afl and mopt might loose ranks.

nice spreadsheet formulars btw!

**Comment by wideglide** [comment]
I would say that the comparison is more nuanced than how you describe it.

> a fuzzer that is good on the first ones, but very bad on the last ones will be rated a good fuzzer. 

In the aflplusplus experiments (2020-05-20-aflplusplus-2), simple aflplusplus was ranked either 1st, 2nd or 4th depending on the ranking algorithm, but generally it performs quite well across all benchmarks. 

In the other experiments (2020-05-11 for example), if we use the percent coverage as the ranking mechanism, the ranks look like this:
```
honggfuzz      97.018742
entropic       92.400031
aflsmart       90.725944
lafintel       90.410083
afl            90.248494
aflplusplus    88.693044
fastcgs_lm     87.601417
mopt           87.521381
aflfast        87.497333
libfuzzer      87.071347
fairfuzz       81.420302
eclipser       71.055928
```

Entropic gains in ranking, but the difference between afl-smart, lafintel and afl is very insignificant.      Gains on one benchmark typically result in losses in another, which keeps the overall ranking a tight race.  I think what the first chart really shows is that there's a small overall "critical difference" between the first 7 fuzzers and selecting the best fuzzer depends on the benchmark.  For example, if you want to fuzz sqlite3, **don't** use honggfuzz or entropic (1st and 2nd), where they really suffer.

**Comment by lszekeres** [comment]
Thanks @vanhauser-thc and @wideglide for flagging this and for the discussion! Yes, as this discussion shows too, cross-benchmark ranking is tricky :)

Unfortunately, there's no single ultimate ranking method that is objectively better than the others. We currently have [6 different](https://github.com/google/fuzzbench/blob/master/analysis/experiment_results.py#L102-L150) cross-benchmark ranking methods available for custom analysis/reports. Often, they put the fuzzers in the same order. However, when fuzzers perform similarly, or when their per-benchmark ranking is less consistent, the different cross-benchmark ranking algorithms can produce very different orders.

This is why I think it's important to consider not just their order, but also how significant the difference is between them. The critical different diagram visualizes this relatively well, [this is why](https://github.com/google/fuzzbench/issues/219#issuecomment-613760632) we chose it as the default one to show. For example, the critical difference diagram of the [report in question](https://www.fuzzbench.com/reports/2020-05-20-aflplusplus-2/index.html) shows that - according to the used statistical test - there's no significant difference between these variants when we consider all benchmarks.

This of course doesn't change the fact that in practice we still want pick one to use, and different folks might want to use different methods to pick a winner. I can think of the following options to support this better with the [reports we publish](https://www.fuzzbench.com/reports):

1) If there's consensus that e.g., the [average normalized score](https://github.com/google/fuzzbench/blob/5c892b4f3b59a1d181f884ad160dc1f94941c01c/analysis/experiment_results.py#L128) is a more useful default, we can change it. One disadvantage of this is that we won't have a cross-benchmark test for statistical significance.

2) We can include multiple ranking results in the published reports. I added such report type in https://github.com/google/fuzzbench/pull/373.

3) We keep one ranking in the published report, but we link to a [more detailed report](https://github.com/google/fuzzbench/issues/267) which includes all the other options and/or a [colab notebook](https://github.com/google/fuzzbench/issues/268) like @wideglide's for custom manual analysis. (We still need to pick a default though)

What do you all think?

**Comment by vanhauser-thc** [comment]
I think the best is to have maybe 3 different overall ranking schemes. 
the time to reach coverage is a good one IMHO.
rank_by_median_and_average_normalized_score would be a good one.
and the third ... mabye the current one, maybe a count of crashes found, ...

I would also propose to replace benchmarks that are not that useful for comparisons (jsoncpp and zlib for example) - unless there is a future fuzzer who can really make a difference there (e.g. zlib -> afl-smart with a peach pit)

**Comment by jonathanmetzman** [comment]
>The problem with the the current method is that there are some test cases where the difference of the fuzzers is between 0.5 and 2 basic blocks (e.g. jsoncpp). and then there are others where there are very large differences (e.g. libpcap). And a fuzzer that is good on the first ones, but very bad on the last ones will be rated a good fuzzer. Which I think most people would agree is not the correct conclusion. (for the first type there is not an effective difference between the fuzzers, for the the last type it is).

I agree with this.

>I think the best is to have maybe 3 different overall ranking schemes

I think having multiple aggregate rankings is a good idea because I agree with @lszekeres that "there's no single ultimate ranking method that is objectively better than the others". Especially because people have different use cases. For example, for us in ClusterFuzz it's not really an improvement for us to add a fuzzer that is best overall but only covers code that was covered by AFL and libFuzzer since we already use them.

On the other hand, I think there is some truth to "this ranking algorithm basically gives the wrong advice to people who want to choose a good fuzzer". I think we should probably determine which ranking best answers the question: "What is the one fuzzer I can use to find the most possible bugs" and make this the most prominent ranking. Even though the current ranking IMO hasn't really been giving strong opinions on which fuzzer is best (since it thinks most differences are not very significant), I wonder if people tend to ignore this.

>I would also propose to replace benchmarks that are not that useful for comparisons (jsoncpp and zlib for example) - unless there is a future fuzzer who can really make a difference there (e.g. zlib -> afl-smart with a peach pit)

I've been considering this as well (see my comment [here](https://github.com/google/fuzzbench/issues/48#issuecomment-627754082)). On some level I think it's biasing the results to eliminate benchmarks with small-to-no differences between fuzzers though.  If someone is trying to find the best fuzzer to use, only using benchmarks where there are differences overstates how much of an advantage the best fuzzer offers. But I think it's a problem that a few benchmarks basically give no actionable information to fuzzer devs because they can't be improved on.

**Comment by jonathanmetzman** [comment]
>If there's consensus that e.g., the average normalized score is a more useful default, we can change it. One disadvantage of this is that we won't have a cross-benchmark test for statistical significance.

I don't know enough about stats to comment, but why is this case? Is it simply that average normalized coverage can't be plugged into the significance tests we use or can this metric not be 
tested for significance in principle.

>What do you all think?

I think I support 1 and 2. Including multiple rankings but changing the first one to be average normalized score.

**Comment by mboehme** [comment]
> I think what the first chart really shows is that there's a small overall "critical difference" between the first 7 fuzzers and selecting the best fuzzer depends on the benchmark. For example, if you want to fuzz sqlite3, don't use honggfuzz or entropic (1st and 2nd), where they really suffer.

This is how fuzzbench seems to be designed now. The overall ranking doesn't seem to matter as much as the individual ranking for different subjects. The subjects are cherry-picked because they exhibit interesting properties, and evidently demonstrate differences across fuzzers. See https://github.com/google/fuzzbench/issues/48#issuecomment-627881407

> I would also propose to replace benchmarks that are not that useful for comparisons

The problem here is that a new fuzzer might suddenly demonstrate a huge difference, but you would not detect it.

> I think we should probably determine which ranking best answers the question: "What is the one fuzzer I can use to find the most possible bugs" and make this the most prominent ranking.

In this case, you probably want to choose a representative set of benchmarks, e.g., by choosing random subjects from OSS-Fuzz. See https://github.com/google/fuzzbench/issues/48#issuecomment-627881407

>> What do you all think?

> I think I support 1 and 2. Including multiple rankings but changing the first one to be average normalized score.

I also support 1 and 2 (and 3 if adding a link to the Google Colab is quickly done. Also it provides more details for the interested user).

**Comment by vanhauser-thc** [comment]
> > I would also propose to replace benchmarks that are not that useful for comparisons
> The problem here is that a new fuzzer might suddenly demonstrate a huge difference, but you would not detect it.

I agree.
However how about if a new fuzzer is added a standalone test run is performed over all targets?
And if a normally not active benchmark shows very different results then adding it to the default target list?

the more targets are benchmarked where there is no useful differences in the results of the fuzzers, the more similar and closer the overall ranking is.

So I would not remove targets, just dont add those to the default run that give not much value and just water the results down.
But having an option to add them again if a fuzzers makes a difference.

> > I think we should probably determine which ranking best answers the question: "What is the one fuzzer I can use to find the most possible bugs" and make this the most prominent ranking.

Maybe. I dont know. depending on the bugs that could be a good or bad benchmark I can image.

coverage and time to coverage - I think these are performance indicators that are always useful.

If a target has its bugs in just one part, then a fuzzer with a very bad coverage might find exactly these bugs, fast, and look like the best.

So having that measurement is good, but IMHO I would not make it the primary one.

**Comment by wideglide** [comment]
> the more targets are benchmarked where there is no useful differences in the results of the fuzzers, the more similar and closer the overall ranking is.

I don't believe this is true in the way you think that it is.

Here are [two sets of ranking charts](https://colab.research.google.com/drive/1GIZzpHDa_TuO_zRyLz-7_4Z5EQJJXlyN), the second one removing four benchmarks where there is no useful differences in coverage.  The rankings do change in some cases, but I'm not sure how the results are watered down?  The **% Coverage Ranking** doesn't change at all, and the **Average Rank (median coverage)** only changes ever so slightly (libfuzzer, fairfuzz, and lafintel switch rankings at the bottom).  Is there a fuzzer (or fuzzers) that you believe should be ranked higher based on its performance?

It seems that the **% Coverage Ranking** is a slightly more intuitive, but there's actually less difference shown between fuzzers using this measure. The biggest change in ranking comes for lafintel, which moves from **10th** to **4th**!  Which is a bit problematic as an overall ranking evaluation, because it consistently performs in the bottom half of evaluated fuzzers per benchmark except for four benchmarks.  

This highlights a different evaluation measure that is not currently captured very well.  How consistent is the performance of a fuzzer?  For example if we first rank fuzzers based on their median reached coverage per benchmark (giving equal rankings for ties), then count how many times a fuzzer is in the top half of the rankings, AFL is the **most** consistent performer.  There's only one benchmark where its ranking is > 6th.  (see the updated [Google Sheet](https://docs.google.com/spreadsheets/d/1pYaf4yrLj0VM2wicfQnfF0A0d_H3KIqukIuiLJPEaNY/edit?usp=sharing), tab 2020-05-11)  

Honestly, this just highlights the difference between AFL variants and non-AFL variants.  Honggfuzz and Entropic do quite well on most benchmarks, Honggfuzz especially showing improved coverage, but in the benchmarks where it doesn't perform well, it drops below 50% over the other evaluated fuzzers (usually 8th).

*Using the % Coverage Ranking means that the critical difference diagram cannot be generated (no overall statistical significance diagram/measure).

**Comment by lszekeres** [comment]
Thanks you all for your inputs, this is a great discussion! 

Let me try to summarize some of the touched topics and what we seem to be converging towards.

####  Time to reach given coverage vs coverage reached in given time

> I think the best is to have maybe 3 different overall ranking schemes.
> the time to reach coverage is a good one IMHO.

This is orthogonal to the aggregate (cross-benchmark level) ranking method in question. We can have a report type that shows the "time to reach a given coverage" instead of the "coverage reached in a given time" for each benchmark (it's another way of showing the data we measure). This doesn't affect the cross-benchmark ranking algorithm, only its input: whether they are time or coverage values. (Tracking issue: https://github.com/google/fuzzbench/issues/327 -- contributions are welcome!)

####  Measuring bugs

> rank_by_median_and_average_normalized_score would be a good one.
> and the third ... mabye the current one, maybe a count of crashes found, ...

Measuring bugs is also orthogonal to the aggregate ranking method. To measure bugs, first we need bug benchmarks though. (Tracking at https://github.com/google/fuzzbench/issues/165 -- contributions are welcome!)

> So having that measurement is good, but IMHO I would not make it the primary one.

I agree, bugs should be an additional metric, not a replacement.

####  What benchmarks

> you probably want to choose a representative set of benchmarks, e.g., by choosing random subjects from OSS-Fuzz. 

I agree that we probably shouldn't pick and choose benchmark based on how the current set of fuzzers perform on them. Instead, we should focus on representativeness, and use an as diverse and large set of benchmarks as possible. This is also orthogonal to the aggregate ranking method though, we can continue discussing this in https://github.com/google/fuzzbench/issues/48.

#### Statistical test

> Is it simply that average normalized coverage can't be plugged into the significance tests we use or can this metric not be tested for significance in principle.

Yes, it's because it cannot be plugged into any non-parametric statistical tests for [block designs](https://en.wikipedia.org/wiki/Blocking_(statistics)) (multi-benchmark experiments) that I'm aware of. (Bit more on this [here](https://github.com/google/fuzzbench/issues/219#issuecomment-613760632
).) But i'm not a statistician either, ideas/contributions are welcome here too!

#### Ranking consistency

> This highlights a different evaluation measure that is not currently captured very well. How consistent is the performance of a fuzzer?

Good point. I think this is something what the currently used critical difference diagram captures to some extennt: the more consistent the fuzzers are, their average rank will differ more. When they are all over the place considering per-benchmark ranking they'll end up in the middle with similar average ranks on the aggregate chart. But there might be better ways to capture this, and we are always open to include new alternative ranking methods/visualizations!

#### What to do

Sounds like we are converging to the following solution:

- Show the `rank_by_median_and_average_normalized_score` and the currently used ranking in the default report we publish.

Additionally, we can also link to colab notebook / more detailed report from.

Let me know (e.g. with a thumbs up) if this sounds good to you. Thanks!
